samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -99,14 +108,24 @@ public void setDefaultSerde(Serde<?> serde) {
 
   @Override
   public <M> OutputStream<M> getOutputStream(String streamId, Serde<M> serde) {
+    StreamSpec streamSpec = runner.getStreamSpec(streamId);
     Preconditions.checkNotNull(serde, "serde must not be null for an output stream.");
-    Preconditions.checkState(!outputStreams.containsKey(runner.getStreamSpec(streamId)),
+    Preconditions.checkState(!outputStreams.containsKey(streamSpec),
         "getOutputStream must not be called multiple times with the same streamId: " + streamId);
 
-    StreamSpec streamSpec = runner.getStreamSpec(streamId);
     KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+    if (inputOperators.containsKey(streamSpec)) {
+      InputOperatorSpec inputOperatorSpec = inputOperators.get(streamSpec);
+      Serde keySerde = inputOperatorSpec.getKeySerde();
+      Serde valueSerde = inputOperatorSpec.getValueSerde();
+      Preconditions.checkState(kvSerdes.getKey().equals(keySerde) && kvSerdes.getValue().equals(valueSerde),
+          String.format("Stream %s is being used both as an input and an output stream. Serde in Samza happens at "
+              + "stream level, so the same key and message Serde must be used for both.", streamId));
+    }
+
+    boolean isKeyedOutput = serde instanceof KVSerde;
```

Comment:
Very Minor: s/isKeyedOutput/isKeyed 

Output appears to be redundant given that the class is `OutputStream`